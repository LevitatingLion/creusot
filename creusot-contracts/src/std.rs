mod boxed;
pub mod clone;
pub mod cmp;
mod default;
mod eq;
mod fun;
pub mod iter;
pub mod mem;
mod num;
pub mod option;
mod result;
mod slice;
pub mod vec;

pub use boxed::*;
pub use clone::*;
pub use eq::*;
pub use fun::*;
pub use vec::*;
