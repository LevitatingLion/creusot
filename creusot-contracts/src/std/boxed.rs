use crate as creusot_contracts;
use creusot_contracts_proc::*;

use std::alloc::Allocator;

extern_spec! {
    mod std {
        mod boxed {
            impl<T> Box<T> {
                #[ensures(*result == val)]
                fn new(val: T) -> Self;
            }

            impl<T, A: Allocator> Box<T, A> {
                #[ensures(**self == *result)]
                #[ensures(*^self == ^result)]
                fn as_mut(&mut self) -> &mut T;

                #[ensures(**self == *result)]
                fn as_ref(&self) -> &T;
            }
        }
    }
}
