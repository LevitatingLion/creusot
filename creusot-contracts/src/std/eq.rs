use crate as creusot_contracts;
use creusot_contracts_proc::*;
use std::alloc::Allocator;

use crate::logic::*;

extern_spec! {
    mod std {
        mod cmp {
            trait PartialEq<Rhs> {
                #[ensures(result == (@self == @rhs))]
                fn eq(&self, rhs: &Rhs) -> bool
                where
                    Self_: Model,
                    Rhs: Model<ModelTy = Self_::ModelTy>;

            }
        }
        mod boxed {
            impl<T, A : Allocator> PartialEq<Box<T, A>> for Box<T, A> {
                #[ensures(result == (@self == @rhs))]
                fn eq(&self, rhs: &Self) -> bool
                where
                    T : Model,
                    T: PartialEq<T> + ?Sized,
                    A: Allocator;
            }
        }
    }
}
