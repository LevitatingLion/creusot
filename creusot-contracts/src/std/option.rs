use crate as creusot_contracts;
use crate::{
    std::{default::DefaultSpec, fun::FnOnceSpec},
    Resolve,
};
use creusot_contracts_proc::*;

extern_spec! {
    mod std {
        mod option {
            impl<T> Option<T> {
                #[ensures(result == (*self != None))]
                fn is_some(&self) -> bool;

                #[ensures(result == (*self == None))]
                fn is_none(&self) -> bool;

                #[requires(self != None)]
                #[ensures(Some(result) == self)]
                fn unwrap(self) -> T;

                #[requires(self != None)]
                #[ensures(Some(result) == self)]
                fn expect(self, msg: &str) -> T;

                #[ensures(self == None ==> result == default)]
                #[ensures(self == None || self == Some(result))]
                fn unwrap_or(self, default: T) -> T;

                #[requires(self != None || f.precondition(()))]
                #[ensures(self == None ==> f.postcondition_once((), result))]
                #[ensures(self == None || self == Some(result))]
                fn unwrap_or_else<F>(self, f: F) -> T
                where
                    F: FnOnce() -> T;

                #[ensures(*self == None ==> result == None && ^self == None)]
                #[ensures(
                    *self == None
                    || exists<r: &mut T> result == Some(r) && *self == Some(*r) && ^self == Some(^r)
                )]
                fn as_mut(&mut self) -> Option<&mut T>;

                #[ensures(*self == None ==> result == None)]
                #[ensures(
                    *self == None || exists<r: &T> result == Some(r) && *self == Some(*r)
                )]
                fn as_ref(&self) -> Option<&T>;

                #[ensures(self == None ==> result == Err(err))]
                #[ensures(self == None || exists<t: T> self == Some(t) && result == Ok(t))]
                fn ok_or<E>(self, err: E) -> Result<T, E>;

                #[requires(self != None || err.precondition(()))]
                #[ensures(
                    self == None ==> exists<e: E> err.postcondition_once((), e) && result == Err(e)
                )]
                #[ensures(self == None || exists<t: T> self == Some(t) && result == Ok(t))]
                fn ok_or_else<E, F>(self, err: F) -> Result<T, E>
                where
                    F: FnOnce() -> E;

                #[requires(self == None || exists<t: T> self == Some(t) && f.precondition((t,)))]
                #[ensures(self == None ==> result == None)]
                #[ensures(
                    self == None
                    || exists<t: T, u: U> self == Some(t) && result == Some(u)
                        && f.postcondition_once((t,), u)
                )]
                fn map<U, F>(self, f: F) -> Option<U>
                where
                    F: FnOnce(T) -> U;

                #[requires(self == None || exists<t: T> self == Some(t) && f.precondition((t,)))]
                #[ensures(self == None ==> result == default)]
                #[ensures(
                    self == None
                    || exists<t: T, u: U> self == Some(t) && result == u
                        && f.postcondition_once((t,), u)
                )]
                fn map_or<U, F>(self, default: U, f: F) -> U
                where
                    F: FnOnce(T) -> U;

                #[requires(
                    (self == None && default.precondition(()))
                    || exists<t: T> self == Some(t) && f.precondition((t,))
                )]
                #[ensures(self == None ==> default.postcondition_once((), result))]
                #[ensures(
                    self == None
                    || exists<t: T, u: U> self == Some(t) && result == u
                        && f.postcondition_once((t,), u)
                )]
                fn map_or_else<U, D, F>(self, default: D, f: F) -> U
                where
                    D: FnOnce() -> U,
                    F: FnOnce(T) -> U;

                #[ensures(self == None ==> result == None)]
                #[ensures(self == None || result == optb)]
                fn and<U>(self, optb: Option<U>) -> Option<U>;

                #[requires(self == None || exists<t: T> self == Some(t) && f.precondition((t,)))]
                #[ensures(self == None ==> result == None)]
                #[ensures(
                    self == None
                    || exists<t: T> self == Some(t) && f.postcondition_once((t,), result)
                )]
                fn and_then<U, F>(self, f: F) -> Option<U>
                where
                    F: FnOnce(T) -> Option<U>;

                #[ensures(self == None ==> result == optb)]
                #[ensures(self == None || result == self)]
                fn or(self, optb: Option<T>) -> Option<T>;

                #[requires(self != None || f.precondition(()))]
                #[ensures(self == None ==> f.postcondition_once((), result))]
                #[ensures(self == None || result == self)]
                fn or_else<F>(self, f: F) -> Option<T>
                where
                    F: FnOnce() -> Option<T>;

                #[requires(
                    self == None || exists<r: &T> self == Some(*r) && predicate.precondition((r,))
                )]
                #[ensures(self == None ==> result == None)]
                #[ensures(
                    self == None
                    || exists<r: &T> self == Some(*r) && (
                        (predicate.postcondition_once((r,), false) && result == None)
                        || (predicate.postcondition_once((r,), true) && result == self)
                    )
                )]
                fn filter<P>(self, predicate: P) -> Option<T>
                where
                    P: FnOnce(&T) -> bool;

                #[ensures(result == *self && ^self == None)]
                fn take(&mut self) -> Option<T>;

                #[ensures(result == *self && ^self == Some(value))]
                fn replace(&mut self, value: T) -> Option<T>;

                #[ensures(self == None ==> result == T::default_log())]
                #[ensures(self == None || self == Some(result))]
                fn unwrap_or_default(self) -> T
                where
                    T: DefaultSpec;
            }

            impl<T> Option<&T> {
                #[ensures(self == None ==> result == None)]
                #[ensures(self == None || exists<t: &T> self == Some(t) && result == Some(*t))]
                fn copied(self) -> Option<T>
                where
                    T: Copy;

                #[ensures(self == None ==> result == None)]
                #[ensures(self == None || exists<t: &T> self == Some(t) && result == Some(*t))]
                fn cloned(self) -> Option<T>
                where
                    T: Clone;
            }

            impl<T> Option<&mut T> {
                #[ensures(self == None ==> result == None)]
                #[ensures(
                    self == None
                    || exists<t: &mut T> self == Some(t) && result == Some(*t) && t.resolve()
                )]
                fn copied(self) -> Option<T>
                where
                    T: Copy;

                #[ensures(self == None ==> result == None)]
                #[ensures(
                    self == None
                    || exists<t: &mut T> self == Some(t) && result == Some(*t) && t.resolve()
                )]
                fn cloned(self) -> Option<T>
                where
                    T: Clone;
            }

            impl<T> Option<Option<T>> {
                #[ensures(self == None ==> result == None)]
                #[ensures(self == None || self == Some(result))]
                fn flatten(self) -> Option<T>;
            }
        }
    }
}

impl<T> DefaultSpec for Option<T> {
    #[logic]
    fn default_log() -> Option<T> {
        None
    }
}
