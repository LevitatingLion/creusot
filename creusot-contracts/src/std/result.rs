use crate as creusot_contracts;
use crate::{std::default::DefaultSpec, Resolve};
use creusot_contracts_proc::*;
use std::fmt::Debug;

extern_spec! {
    mod std {
        mod result {
            impl<T, E> Result<T, E> {
                #[ensures(result == exists<t: T> *self == Ok(t))]
                fn is_ok(&self) -> bool;

                #[ensures(result == exists<e: E> *self == Err(e))]
                fn is_err(&self) -> bool;

                #[ensures(forall<t: T> self == Ok(t) ==> result == Some(t))]
                #[ensures((exists<e: E> self == Err(e)) ==> result == None)]
                fn ok(self) -> Option<T>;

                #[ensures((exists<t: T> self == Ok(t)) ==> result == None)]
                #[ensures(forall<e: E> self == Err(e) ==> result == Some(e))]
                fn err(self) -> Option<E>;

                #[ensures(forall<t: &T> *self == Ok(*t) ==> result == Ok(t))]
                #[ensures(forall<e: &E> *self == Err(*e) ==> result == Err(e))]
                #[ensures((exists<t: &T> *self == Ok(*t))|| (exists<e: &E> *self == Err(*e)))]
                fn as_ref(&self) -> Result<&T, &E>;

                #[ensures(
                    forall<t: &mut T> *self == Ok(*t) ==> ^self == Ok(^t) && result == Ok(t)
                )]
                #[ensures(
                    forall<e: &mut E> *self == Err(*e) ==> ^self == Err(^e) && result == Err(e)
                )]
                #[ensures(
                    (exists<t: &mut T> *self == Ok(*t))|| (exists<e: &mut E> *self == Err(*e))
                )]
                fn as_mut(&mut self) -> Result<&mut T, &mut E>;

                #[requires(exists<t: T> self == Ok(t))]
                #[ensures(Ok(result) == self)]
                fn unwrap(self) -> T
                where
                    E: Debug;

                #[requires(exists<t: T> self == Ok(t))]
                #[ensures(Ok(result) == self)]
                fn expect(self, msg: &str) -> T
                where
                    E: Debug;

                #[requires(exists<e: E> self == Err(e))]
                #[ensures(Err(result) == self)]
                fn unwrap_err(self) -> E
                where
                    T: Debug;

                #[ensures(forall<t: T> self == Ok(t) ==> result == t)]
                #[ensures((exists<e: E> self == Err(e)) ==> result == default)]
                fn unwrap_or(self, default: T) -> T;

                #[ensures(forall<t: T> self == Ok(t) ==> result == t)]
                #[ensures((exists<e: E> self == Err(e)) ==> result == T::default_log())]
                fn unwrap_or_default(self) -> T
                where
                    T: DefaultSpec;

                #[ensures((exists<t: T> self == Ok(t)) ==> result == res)]
                #[ensures(forall<e: E> self == Err(e) ==> result == Err(e))]
                fn and<U>(self, res: Result<U, E>) -> Result<U, E>;

                #[ensures(forall<t: T> self == Ok(t) ==> result == Ok(t))]
                #[ensures((exists<e: E> self == Err(e)) ==> result == res)]
                fn or<F>(self, res: Result<T, F>) -> Result<T, F>;
            }

            impl<T, E> Result<&T, E> {
                #[ensures(forall<t: &T> self == Ok(t) ==> result == Ok(*t))]
                #[ensures(forall<e: E> self == Err(e) ==> result == Err(e))]
                fn copied(self) -> Result<T, E>
                where
                    T: Copy;

                #[ensures(forall<t: &T> self == Ok(t) ==> result == Ok(*t))]
                #[ensures(forall<e: E> self == Err(e) ==> result == Err(e))]
                fn cloned(self) -> Result<T, E>
                where
                    T: Clone;
            }

            impl<T, E> Result<&mut T, E> {
                #[ensures(forall<t: &mut T> self == Ok(t) ==> result == Ok(*t) && t.resolve())]
                #[ensures(forall<e: E> self == Err(e) ==> result == Err(e))]
                fn copied(self) -> Result<T, E>
                where
                    T: Copy;

                #[ensures(forall<t: &mut T> self == Ok(t) ==> result == Ok(*t) && t.resolve())]
                #[ensures(forall<e: E> self == Err(e) ==> result == Err(e))]
                fn cloned(self) -> Result<T, E>
                where
                    T: Clone;
            }

            impl<T, E> Result<Option<T>, E> {
                #[ensures(self == Ok(None) ==> result == None)]
                #[ensures(forall<t: T> self == Ok(Some(t)) ==> result == Some(Ok(t)))]
                #[ensures(forall<e: E> self == Err(e) ==> result == Some(Err(e)))]
                fn transpose(self) -> Option<Result<T, E>>;
            }
        }
    }
}
