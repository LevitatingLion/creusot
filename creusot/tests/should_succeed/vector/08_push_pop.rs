// WHY3PROVE Z3
extern crate creusot_contracts;

use creusot_contracts::std::*;
use creusot_contracts::*;


fn vec_push_pop<T>(v: &mut Vec<T>, e: T) {
    let old_v = Ghost::record(&v);

    v.push(e);
    let r = v.pop();

    match r {
      Some(v) => (),
      None => (),
    };

    proof_assert!(@v == @@old_v);
    // Unprovable (automatically at least) without the previous line
    proof_assert!(@v === @@old_v);
}