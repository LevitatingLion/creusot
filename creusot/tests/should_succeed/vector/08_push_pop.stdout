module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type creusotcontracts_std1_vec_vec 't  
  type creusotcontracts_logic_ghost_ghost 't = 
    | CreusotContracts_Logic_Ghost_Ghost opaque_ptr
    
  type core_option_option 't = 
    | Core_Option_Option_None
    | Core_Option_Option_Some 't
    
  function core_option_option_Some_0 (self : core_option_option 't) : 't
  val core_option_option_Some_0 (self : core_option_option 't) : 't
    ensures { result = core_option_option_Some_0 self }
    
  axiom core_option_option_Some_0_acc : forall a : 't . core_option_option_Some_0 (Core_Option_Option_Some a : core_option_option 't) = a
end
module CreusotContracts_Logic_Model_Model_ModelTy
  type self   
  type modelTy   
end
module CreusotContracts_Logic_Model_Model_Model_Interface
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Model_Model
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl1_Model_Interface
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  function model (self : borrowed t) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl1_Model
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  clone CreusotContracts_Logic_Model_Model_Model_Interface as Model0 with type self = t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  function model (self : borrowed t) : ModelTy0.modelTy = 
    Model0.model ( * self)
end
module CreusotContracts_Logic_Ghost_Impl0_Model_Interface
  type t   
  use Type
  function model (self : Type.creusotcontracts_logic_ghost_ghost t) : t
end
module CreusotContracts_Logic_Ghost_Impl0_Model
  type t   
  use Type
  function model (self : Type.creusotcontracts_logic_ghost_ghost t) : t
end
module CreusotContracts_Logic_Model_Impl1_ModelTy
  type t   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  type modelTy  = 
    ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl1
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy2 with type self = t
  clone CreusotContracts_Logic_Model_Model_Model as Model2 with type self = t, type ModelTy0.modelTy = ModelTy2.modelTy
  clone CreusotContracts_Logic_Model_Impl1_Model as Model0 with type t = t, type ModelTy0.modelTy = ModelTy2.modelTy,
  function Model0.model = Model2.model
  clone CreusotContracts_Logic_Model_Impl1_ModelTy as ModelTy0 with type t = t, type ModelTy0.modelTy = ModelTy2.modelTy
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = borrowed t, function model = Model0.model,
  type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = borrowed t,
  type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Logic_Ghost_Impl0_ModelTy
  type t   
  type modelTy  = 
    t
end
module CreusotContracts_Logic_Ghost_Impl0
  type t   
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model as Model0 with type t = t
  clone CreusotContracts_Logic_Ghost_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = Type.creusotcontracts_logic_ghost_ghost t,
  function model = Model0.model, type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = Type.creusotcontracts_logic_ghost_ghost t,
  type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Std1_Vec_Impl0_ModelTy
  type t   
  use seq.Seq
  type modelTy  = 
    Seq.seq t
end
module CreusotContracts_Std1_Vec_Impl0_Model_Interface
  type t   
  use Type
  use seq.Seq
  function model (self : Type.creusotcontracts_std1_vec_vec t) : Seq.seq t
end
module CreusotContracts_Std1_Vec_Impl0_Model
  type t   
  use Type
  use seq.Seq
  function model (self : Type.creusotcontracts_std1_vec_vec t) : Seq.seq t
end
module CreusotContracts_Std1_Vec_Impl0
  type t   
  use Type
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model0 with type t = t
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = Type.creusotcontracts_std1_vec_vec t,
  function model = Model0.model, type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = Type.creusotcontracts_std1_vec_vec t,
  type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Ghost_Impl1_Record_Interface
  type t   
  use prelude.Prelude
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model_Interface as Model0 with type t = t
  val record [@cfg:stackify] (a : t) : Type.creusotcontracts_logic_ghost_ghost t
    ensures { Model0.model result = a }
    
end
module CreusotContracts_Logic_Ghost_Impl1_Record
  type t   
  use prelude.Prelude
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model_Interface as Model0 with type t = t
  val record [@cfg:stackify] (a : t) : Type.creusotcontracts_logic_ghost_ghost t
    ensures { Model0.model result = a }
    
end
module CreusotContracts_Std1_Vec_Impl1_Push_Interface
  type t   
  use seq.Seq
  use prelude.Prelude
  use Type
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl1_Model_Interface as Model1 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t
  val push [@cfg:stackify] (self : borrowed (Type.creusotcontracts_std1_vec_vec t)) (v : t) : ()
    ensures { Model0.model ( ^ self) = Seq.snoc (Model1.model self) v }
    
end
module CreusotContracts_Std1_Vec_Impl1_Push
  type t   
  use seq.Seq
  use prelude.Prelude
  use Type
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl1_Model_Interface as Model1 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t
  val push [@cfg:stackify] (self : borrowed (Type.creusotcontracts_std1_vec_vec t)) (v : t) : ()
    ensures { Model0.model ( ^ self) = Seq.snoc (Model1.model self) v }
    
end
module CreusotContracts_Std1_Vec_Impl1_Pop_Interface
  type t   
  use Type
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model1 with type t = t
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl1_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val pop [@cfg:stackify] (self : borrowed (Type.creusotcontracts_std1_vec_vec t)) : Type.core_option_option t
    ensures { match (result) with
      | Type.Core_Option_Option_Some t -> Model0.model self = Seq.snoc (Model1.model ( ^ self)) t
      | Type.Core_Option_Option_None -> Seq.length (Model0.model self) = Seq.length (Model1.model ( ^ self)) && Seq.length (Model0.model self) = 0
      end }
    
end
module CreusotContracts_Std1_Vec_Impl1_Pop
  type t   
  use Type
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model1 with type t = t
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl1_Model_Interface as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val pop [@cfg:stackify] (self : borrowed (Type.creusotcontracts_std1_vec_vec t)) : Type.core_option_option t
    ensures { match (result) with
      | Type.Core_Option_Option_Some t -> Model0.model self = Seq.snoc (Model1.model ( ^ self)) t
      | Type.Core_Option_Option_None -> Seq.length (Model0.model self) = Seq.length (Model1.model ( ^ self)) && Seq.length (Model0.model self) = 0
      end }
    
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve_Interface
  type t   
  use prelude.Prelude
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Logic_Resolve_Impl1_Resolve
  type t   
  use prelude.Prelude
  predicate resolve (self : borrowed t) = 
     ^ self =  * self
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve_Interface
  type t   
  predicate resolve (self : t)
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve
  type t   
  predicate resolve (self : t) = 
    true
end
module CreusotContracts_Logic_Resolve_Impl1
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = borrowed t,
  predicate resolve = Resolve0.resolve
end
module CreusotContracts_Logic_Resolve_Impl2
  type t   
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve0 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = t,
  predicate resolve = Resolve0.resolve
end
module C08PushPop_VecPushPop_Interface
  type t   
  use prelude.Prelude
  use Type
  val vec_push_pop [@cfg:stackify] (v : borrowed (Type.creusotcontracts_std1_vec_vec t)) (e : t) : ()
end
module C08PushPop_VecPushPop
  type t   
  use seq.Seq
  use prelude.Prelude
  use Type
  clone CreusotContracts_Logic_Ghost_Impl0_Model as Model1 with type t = borrowed (Type.creusotcontracts_std1_vec_vec t)
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model2 with type t = t
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl1_Model as Model0 with type t = Type.creusotcontracts_std1_vec_vec t,
  type ModelTy0.modelTy = ModelTy0.modelTy, function Model0.model = Model2.model
  use mach.int.Int
  use mach.int.Int64
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve5 with type t = ()
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve4 with type self = Type.core_option_option t
  clone CreusotContracts_Logic_Resolve_Impl1_Resolve as Resolve3 with type t = Type.creusotcontracts_std1_vec_vec t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve2 with type self = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = Type.creusotcontracts_logic_ghost_ghost (borrowed (Type.creusotcontracts_std1_vec_vec t))
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve0 with type self = borrowed (Type.creusotcontracts_std1_vec_vec t)
  clone CreusotContracts_Logic_Ghost_Impl1_Record_Interface as Record0 with type t = borrowed (Type.creusotcontracts_std1_vec_vec t),
  function Model0.model = Model1.model
  clone CreusotContracts_Std1_Vec_Impl1_Pop_Interface as Pop0 with type t = t, function Model0.model = Model0.model,
  function Model1.model = Model2.model
  clone CreusotContracts_Std1_Vec_Impl1_Push_Interface as Push0 with type t = t, function Model0.model = Model2.model,
  function Model1.model = Model0.model
  let rec cfg vec_push_pop [@cfg:stackify] (v : borrowed (Type.creusotcontracts_std1_vec_vec t)) (e : t) : () = 
  var _0 : ();
  var v_1 : borrowed (Type.creusotcontracts_std1_vec_vec t);
  var e_2 : t;
  var old_v_3 : Type.creusotcontracts_logic_ghost_ghost (borrowed (Type.creusotcontracts_std1_vec_vec t));
  var _4 : borrowed (Type.creusotcontracts_std1_vec_vec t);
  var _5 : borrowed (Type.creusotcontracts_std1_vec_vec t);
  var _6 : ();
  var _7 : borrowed (Type.creusotcontracts_std1_vec_vec t);
  var _8 : t;
  var r_9 : Type.core_option_option t;
  var _10 : borrowed (Type.creusotcontracts_std1_vec_vec t);
  var _11 : ();
  var _12 : isize;
  var v_13 : t;
  var _14 : ();
  var _15 : ();
  {
    v_1 <- v;
    e_2 <- e;
    goto BB0
  }
  BB0 {
    _5 <- v_1;
    _4 <- _5;
    assume { Resolve0.resolve _5 };
    old_v_3 <- Record0.record _4;
    goto BB1
  }
  BB1 {
    assume { Resolve1.resolve old_v_3 };
    _7 <- borrow_mut ( * v_1);
    v_1 <- { v_1 with current = ( ^ _7) };
    assume { Resolve2.resolve _8 };
    _8 <- e_2;
    _6 <- Push0.push _7 _8;
    goto BB2
  }
  BB2 {
    _10 <- borrow_mut ( * v_1);
    v_1 <- { v_1 with current = ( ^ _10) };
    r_9 <- Pop0.pop _10;
    goto BB3
  }
  BB3 {
    assume { Resolve3.resolve v_1 };
    switch (r_9)
      | Type.Core_Option_Option_None -> goto BB4
      | Type.Core_Option_Option_Some _ -> goto BB6
      end
  }
  BB4 {
    _11 <- ();
    assume { Resolve5.resolve _11 };
    goto BB8
  }
  BB5 {
    assume { Resolve4.resolve r_9 };
    absurd
  }
  BB6 {
    assume { Resolve2.resolve v_13 };
    v_13 <- Type.core_option_option_Some_0 r_9;
    _11 <- ();
    assume { Resolve5.resolve _11 };
    goto BB7
  }
  BB7 {
    assume { Resolve2.resolve v_13 };
    goto BB8
  }
  BB8 {
    assert { Seq.(==) (Model0.model v_1) (Model0.model (Model1.model old_v_3)) };
    _14 <- ();
    assume { Resolve5.resolve _14 };
    assert { Model0.model v_1 = Model0.model (Model1.model old_v_3) };
    _15 <- ();
    assume { Resolve5.resolve _15 };
    _0 <- ();
    goto BB9
  }
  BB9 {
    assume { Resolve4.resolve r_9 };
    goto BB10
  }
  BB10 {
    return _0
  }
  
end
